### PR TITLE
generate and exchange curve security keys on the fly

### PR DESCRIPTION
**Pull request by garlick**
Do away with the need to run `flux keygen` before running a flux insatnce.

This PR generates a unique CURVE key pair per broker, in memory, and shares the public keys using PMI, as proposed in #2767 .  It works when flux launches flux, or when a foreign resource manager launches flux.

The system instance continues to use keys generated by `flux keygen`, although the key directory layout changes somewhat so that keys will need to be regenerated.  (There is only one certificate stored in the file $HOME/.flux/curve now.  Formerly that was a directory containing 4 files, representing 2 certs).

I haven't tested the performance impact of this change.  It does increase the amount of traffic in the PMI exchange such that every broker now contributes a "business card", and a broker fetches the business cards of its parent and all children.  The business card contains the endpoint URI and the public key.  That's a bit more PMI activity than before though I'm hoping it doesn't add much.

Still todo: it should be possible to skip the PMI exchange of public keys if the enclosing instance is running as the same user.  In that case the enclosing instance should be able to share a single cert wtih all brokers, improving startup time for the workflow case.

### CODE REVIEW

**Comment by grondo** [comment]
So all these keys I've collected over the years are now worthless?

**Comment by garlick** [comment]
Force pushed with some changes
* rebased on current master
* skip PMI put, barriers on size=1 instance
* add support for config file pointing to certificate for system instance
* change `flux keygen` so it just has one positional argument, an output file
* add tests for config file boot with configured key
* update docs
* cull some code that's now obsolete/unused

I'm thinking that the optimization alluded to above where enclosing instance provides keys if running as the same user could come in another PR.  This PR is fairly self contained.

Still need to check performance.

**Comment by garlick** [comment]
BTW, there is not much reason to have the "zsecurity" class after this PR is merged. I did remove it in my development branch, but found myself adding a lot of extra code to recreating testing against the broker that's already implemented in the zsecurity unit test.  I conclude that it should probably be left alone for now.

**Comment by garlick** [comment]
Just forced a push with
* zsecurity class removed
* broker overlay unit test added (covers what zsecurity test used to cover)
* use an in-memory `zcertstore_t` instead of on-disk (for authorized public keys)
* rebase on current master

Now unfortunately this PR contains fixes to zsecurity and its eventual removal so I'll need to figure out how to refactor to avoid the incremental development.

**Comment by garlick** [comment]
> use an in-memory zcertstore_t instead of on-disk (for authorized public keys)

Aww, this doesn't work on centos7, since support was only just added to czmq in 2016 :-(

**Comment by grondo** [comment]
> Aww, this doesn't work on centos7, since support was only just added to czmq in 2016 :-(

Could we pull it in via a libmissing type thing?

**Comment by garlick** [comment]
> Could we pull it in via a libmissing type thing?

That may be the path of least resistance.   Should I just vendor libczmq?  One less dependency, and it doesn't change much.

**Comment by garlick** [comment]
I think I'll see if I can implement a simple ZAP handler in the broker thread and skip pulling anything in.

**Comment by garlick** [comment]
Yep, that worked out really well actually.  OK, now to refactor this mess of a PR.

**Comment by garlick** [comment]
I've removed the WIP as I think this is getting close.

To avoid breaking sched (and git bisect), I made the new PATH argument to `flux-keygen` optional.   It prints a warning if the PATH is unspecified.  Hopefully that's OK - it's only for sys admins now.

Sorry about the rather long commits.  I had things nicely broken down before, but in clearing up the incremental development and ultimate removal of the zsecuriity class, I had to squash some of that down.

I'm fairly confident that this works as advertised, and that Flux is more secure than it was before, and also more convenient.  We are only exchanging public keys over PMI, and now neither secret nor public keys ever reach disk (except for the system instance, where they can be more readily protected).  Their lifetime and scope are also limited to a single instance, instead of being shared across all instances run by the same user.

Discontinuing use of the non-portable `zauth` class from CZMQ not only permits authorized public keys to be maintained in memory only, but eliminates the "actor thread" and utilizes the flux reactor for zeromq authentication processing.  That fact opens the door to implementing GSSAPI authentication, which is not supported by even modern versions of zauth.  We can also log authentication attempts, and make any failed ones stand out at LOG_ERR level.

**Comment by grondo** [comment]
Wow, this turned out to be a huge improvement! I'll give the latest a whirl on fluke a bit later today.

**Comment by garlick** [comment]
Thanks!

**Comment by grondo** [comment]
I ran a few different sizes of Flux instances within a 94 node flux job on fluke and timed `flux mini submit -n N flux start flux getattr size` on this branch vs current master:

The first set of times are for this branch, the second for master. It doesn't appear there's much difference in this case!

```
NTASKS RUNTIME
 376 10.48s
 256 9.492s
 128 8.488s
  64 7.801s
  32 6.467s
  16 5.380s
   8 4.570s
   4 3.567s
   2 2.745s

NTASKS RUNTIME
 376 10.81s
 256 9.498s
 128 8.301s
  64 7.422s
  32 6.537s
  16 5.423s
   8 4.676s
   4 3.568s
   2 3.482s
```

For some reason running the flux-start is slower directly under srun

**Comment by garlick** [comment]
Awesome, thanks for that testing!

**Comment by garlick** [comment]
Just tacked on a bit more test coverage and rebased on current master - hope that's OK.  Test coverage is really close to the diff threshold so hopefully this will put it over the edge?

**Comment by garlick** [comment]
So close!  One more try.

**Comment by garlick** [comment]
Well that's got it.  Finally!

**Comment by grondo** [comment]
Nice work! I'll try to give a cursory code review but given that this works well (and I ran it through a battery of tests on flux) and is a huge improvement for users, I'm inclined to approve immediately.

**Comment by garlick** [comment]
Hang on, just noticed a small error. Let me push one more time.

File: doc/man1/flux-keygen.rst [hunk 1]
```diff
@@ -8,46 +8,26 @@ flux-keygen(1)
...
+**flux** **keygen** *PATH*
 
 
 DESCRIPTION
 ===========
 
-flux-keygen(1) generates long-term keys for Flux security.
-Keys are written to files in *$HOME/.flux*.
+flux-keygen(1) generates a long-term CURVE certificate used to secure
+a Flux system instance's overlay network.
```
Comment (grondo) [inline doc/man1/flux-keygen.rst:18]:
```suggestion
flux-keygen(1) generates a long-term CURVE certificate used to secure
the overlay network of a Flux system instance.
```

File: doc/man1/flux-keygen.rst [hunk 2]
```diff
@@ -8,46 +8,26 @@ flux-keygen(1)
...
-=======
-
-**-f, --force**
-   Generate new keys, even if they already exist.
-
-**-p, --plain**
-   Generate PLAIN key.
+Flux instances that bootstrap with PMI do not require a configured certificate.
+In that case, each broker self-generates a unique certificate in and the
+public keys are exchanged with PMI.
```
Comment (grondo) [inline doc/man1/flux-keygen.rst:30]:
```suggestion
In that case, each broker self-generates a unique certificate and the
public keys are exchanged with PMI.
```

File: doc/man1/flux-keygen.rst (hunk 1 above)
Comment (grondo) [reply doc/man1/flux-keygen.rst:18]: oops meant to start a review, not post a single comment, D'oh.

File: doc/man1/flux-keygen.rst [hunk 3]
```diff
@@ -8,46 +8,26 @@ flux-keygen(1)
...
 ===========
 
-flux-keygen(1) generates long-term keys for Flux security.
-Keys are written to files in *$HOME/.flux*.
+flux-keygen(1) generates a long-term CURVE certificate used to secure
+a Flux system instance's overlay network.
 
-Flux comms sessions implement cryptographic privacy and data integrity
-when data is sent over a network. Point to point ZeroMQ TCP connections
+The flux overlay network implements cryptographic privacy and data integrity
```
Comment (grondo) [inline doc/man1/flux-keygen.rst:20]: For consistency of capitalization:

```suggestion
The Flux overlay network implements cryptographic privacy and data integrity
```

**Review by grondo — APPROVED** [review]
Looks good to me! Just some trivial doc suggestions you can take or leave.

I think you fixed at least part of #3378 in this rework also (not the bind to ipv6 port, but the mangling of ipv6 zone local addresses with `%h` in them)

**Comment by grondo** [comment]
I restarted the asan builder which timed out.

**Comment by garlick** [comment]
Thanks for the doc feedback.  I just fixed those as suggested, rebased on current master, and force pushed.

**Pull request merged**